core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java

```diff
@@ -87,27 +85,37 @@
     private BuildServiceConfig config;
     private RegistryService.RegistryConfig registryConfig;
     private AuthConfigFactory authConfigFactory;
+    private BinaryInputArchiveBuilder binaryInputArchiveBuilder;
 
 
-    public OpenshiftBuildService(OpenShiftClient client, Logger log, ServiceHub dockerServiceHub, BuildServiceConfig config) {
+    public OpenshiftBuildService(OpenShiftClient client, Logger log, ServiceHub dockerServiceHub, BuildServiceConfig config, BinaryInputArchiveBuilder binaryInputArchiveBuilder) {
         Objects.requireNonNull(client, "client");
         Objects.requireNonNull(log, "log");
         Objects.requireNonNull(dockerServiceHub, "dockerServiceHub");
         Objects.requireNonNull(config, "config");
+        Objects.requireNonNull(binaryInputArchiveBuilder, "binaryInputArchiveBuilder");
 
         this.client = client;
         this.log = log;
         this.dockerServiceHub = dockerServiceHub;
         this.config = config;
+        this.binaryInputArchiveBuilder = binaryInputArchiveBuilder;
     }
 
     @Override
     public void build(ImageConfiguration imageConfig) throws Fabric8ServiceException {
         String buildName = null;
         try {
-            ImageName imageName = new ImageName(imageConfig.getName());
+            // Create an empty directory
+            File targetDir = createEmptyTargetDir();
+
+            // Fill the directory with the data in the right directory structure
+            binaryInputArchiveBuilder.createBinaryInput(dockerServiceHub, targetDir, imageConfig);
```

Comment:
What's about adding a `File` return value to the `BinaryInputArchiveBuilder` which then holds the the TAR archive ? 

Advantages:

* Get rid of `createTar()` method
* Simplify interface to one method
* No need for the `targetDir` parameter. This just would be an implementation details.

If you don't mind I would change it like this and push it to your branch (just to speed up things and to clarify my point).

@rohanKanojia would this be ok for you ?